content/blog/2022-05-22-object/index.md

```diff
@@ -0,0 +1,162 @@
+
+## Syntax
+Now let's look at the syntax in the example program.
+
+### Class Definition / This
+A class definition looks somewhat similar to a function definition, with three major differences:
+1. We could now define functions in the scope of class.
+2. We can define variables without taking any operations. (As we are not running anything.)
+3. We now have `this` identifier to indicate fields/functions within the class.
+
+Compared to the OOP implementation in a high-level language, there's currently no inheritance, no encapsulation (every field can be visited globally by the `dot` operator), no polymorphism (this abstraction is too high-level and would cause confusion, which makes no sense implementing in Bril). 
+
+### Reference / `dot` / `get` / `set`
+`Reference` is a new type which is analogous to the `Pointer` type, but with a higher level of abstraction as it deals with objects rather than heap memory. We use `dot` to obtain the reference to each fields of object, including any class functions. The `dot` operation takes in two arguments, which are object names and field names --- `dot rect1 area` just means `rect1.area`. A reference to a class function can be directly passed into `call` operation, while a reference to a field variable must be used with `get` and `set`, which are similar to `load` and `store` in pointer operations. I was very hestitated in adding `get` and `set` since it might make the syntax too complicated, however, I believe it's useful in following ways:
+1. Just like `dot` operator, `get` and `set` are very common abstractions in object.
+2. `get` and `set` could hide the underlying implementation of class (pointer operations) which we translate into.
+3. `get` and `set` is independent of type and implementation.
+
+### New / Delete
+`new` allocates the space for an object with regard to a certain class. It also statically bound the object to that class. `delete` frees up the space for the object.
+
+## Translation
+None of the extended syntax is perserved into the translated Bril program. As I mentioned earlier, I am trying to be as being non-invasive as possible to keep Bril simple and low-level. Therefore, I did not change the Bril Interpreter even though it was very tempting to me. The implementation of this work is therefore a translation pass from extended syntax to the core Bril syntax.
+
+The underlying implementation of all fields are pointers, as an object is stored in heap. My initial attempt was to follow a similar approach as `high level language -> LLVM`. It does not work well in this case, because the tools I have is more restrictive and limited. The pointers are strictly typed and are limited to certain pointer arithmetic. So that I was unable to put all the fields in a class into a consecutive block of heap. Instead, I have to allocate a block per field. Hence a `new` operation is essentially `alloc` on every fields, and `delete` is just `free` for every fields.
```

Comment:
Probably don't use code backticks around "high level language -> LLVM." I would use a normal natural-language phrase, such as "as compiling a high-level language to LLVM."